crates/bevy_reflect/src/struct_trait.rs

```diff
@@ -73,6 +73,7 @@ pub struct StructInfo {
     type_name: &'static str,
     type_id: TypeId,
     fields: Box<[NamedField]>,
+    field_names: Box<[&'static str]>,
```

Comment:
@cart I know we decided against this in [this](https://github.com/bevyengine/bevy/pull/5723#discussion_r960008538) previous comment, but I discovered that the `deserializer.deserialize_struct` hint is actually used at least by `bincode` to determine how many fields exist in struct-like types.

If we don't want to support `bincode` we could just remove it.

Alternatively, we could attempt to split the logic based on `deserializer.is_human_readable` or something. Then if we're deserializing a non-self-describing format (such as `bincode`), we manually deserialize it as a tuple like `deserializer.deserialize_tuple(struct_info.field_len(), StructVisitor{...})`. And if it is human-readable, we just handle it as we currently do.

One concern with that is there might still be a difference between how certain formats deserialize. I actually ran into an issue where `rmp-serde` deserializes variant names as identifiers while `bincode` deserializes them as integers. This meant a simple `is_human_readable` couldn't actually handle both cases and a brand new `VariantVisitor` had to be made.

So any thoughts on this with the new context? Keep this change? Replace it with the human-readable check? Something else?

Reply:
My variant name issue might actually be related to [this](https://github.com/bevyengine/bevy/pull/6140?notification_referrer_id=NT_kwDOAvf-ibM0NTE5MzM1NDQ2OjQ5ODA2OTg1#issuecomment-1265843491) comment. So the split might not be as bad as originally thought.

The only other concern would be making that file a lot more complicated. But I think that can be addressed later with a good ol' refactor.

Reply:
I'm fine to wait to refactor after.